### PR TITLE
Fix build warnings

### PR DESCRIPTION
**Pull request by ghost**
Cleans up a whole bunch. Remaining offenders are:
- Parsers.pyx and tokenizer. Signed/unsigned hell, made worse by python/cython/c mixing.
- interval.pyx, not touched because large PR in progress in #27100.
- np_datetime.c and np_datetime_strings.c. These are linked into 15+ cython modules as extra object files, but they '#include <datetime.h>`, which expects only standalone modules, who properly init to do it. Kinda painful to fix, though it means each function is compiled and kept in memory 15 times.

### CODE REVIEW

File: pandas/_libs/groupby_helper.pxi.in [hunk 1]
```diff
@@ -76,7 +76,11 @@ def group_last_{{name}}({{c_type}}[:, :] out,
                 val = values[i, j]
 
                 # not nan
-                if val == val and val != {{nan_val}}:
+                if (
+                {{if not name.startswith("int")}}
```
Comment (WillAyd) [inline pandas/_libs/groupby_helper.pxi.in]: This is a little hacky - IIRC in other modules we template of the NA value. Is that not an option instead?

File: pandas/_libs/hashtable.pxd [hunk 2]
```diff
@@ -41,7 +41,7 @@ cdef class StringHashTable(HashTable):
 
 cdef struct Int64VectorData:
     int64_t *data
-    size_t n, m
+    ssize_t n, m
```
Comment (WillAyd) [inline pandas/_libs/hashtable.pxd]: Not sure about using ssize_t - does Py_size_t not work here instead?

**Review by WillAyd — CHANGES_REQUESTED** [review]
Brief review from phone can dive deep later. Good to take a look and clean these up for sure. 

Can you post the warnings that you are resolving?

File: pandas/_libs/groupby_helper.pxi.in (hunk 1 above)
Comment (pilkibun) [reply pandas/_libs/groupby_helper.pxi.in]: That's not the warning. The issue is that (val == val) is always true for ints, and the compiler flags it as a possible mistake (since why would you?). this omits this part of the check for `ints`.

Comment (jreback) [reply pandas/_libs/groupby_helper.pxi.in]: we don’t do this anywhere else
so rather have this consistently done instead of we decide (separate pr)

**Review by jreback — CHANGES_REQUESTED** [review]
comment

File: pandas/_libs/hashtable.pxd (hunk 2 above)
Comment (pilkibun) [reply pandas/_libs/hashtable.pxd]: no, since one is signed and the other unsigned. look at
https://github.com/pandas-dev/pandas/blob/d054c037132b6cdb61cdf3ed90499287f163aeb4/pandas/_libs/hashtable_class_helper.pxi.in#L32-L34

The two are supposed to match, no ? ssize_t is good. it avoids undeflows in calculations, and the lost bit doesn't matter even on 32bit archs.

File: pandas/_libs/groupby_helper.pxi.in (hunk 1 above)
Comment (pilkibun) [reply pandas/_libs/groupby_helper.pxi.in]: don't understand.

File: pandas/_libs/groupby.pyx [hunk 3]
```diff
@@ -260,7 +260,7 @@ def group_shift_indexer(int64_t[:] out, const int64_t[:] labels,
                         int ngroups, int periods):
     cdef:
         Py_ssize_t N, i, j, ii
-        int offset, sign
+        int offset = 0, sign
```
Comment (pilkibun) [inline pandas/_libs/groupby.pyx]: possible uninitialized use, probably when periods=0. I don't see it, and I think it's a false positive.

File: pandas/_libs/hashtable.pyx [hunk 4]
```diff
@@ -44,7 +44,7 @@ cdef int64_t NPY_NAT = util.get_nat()
 _SIZE_HINT_LIMIT = (1 << 20) + 7
 
 
-cdef size_t _INIT_VEC_CAP = 128
+cdef ssize_t _INIT_VEC_CAP = 128
```
Comment (pilkibun) [inline pandas/_libs/hashtable.pyx]: ssize_t all the way down, to match above.

File: pandas/_libs/hashtable_func_helper.pxi.in [hunk 5]
```diff
@@ -241,7 +241,7 @@ def ismember_{{dtype}}({{scalar}}[:] arr, {{scalar}}[:] values):
 
     # construct the table
     n = len(values)
-    kh_resize_{{ttype}}(table, min(n, len(values)))
+    kh_resize_{{ttype}}(table, n)
```
Comment (pilkibun) [inline pandas/_libs/hashtable_func_helper.pxi.in]: dead code, and warning about min comparing signed and unsigned. fixing one fixes the other.

File: pandas/_libs/index.pyx [hunk 6]
```diff
@@ -352,7 +352,7 @@ cdef class IndexEngine:
 
 cdef Py_ssize_t _bin_search(ndarray values, object val) except -1:
     cdef:
-        Py_ssize_t mid, lo = 0, hi = len(values) - 1
+        Py_ssize_t mid = 0, lo = 0, hi = len(values) - 1
```
Comment (pilkibun) [inline pandas/_libs/index.pyx]: This is a real bug I think. mid is referenced in the return value, if len(values)=1 so the main loop immediately falls through.

File: pandas/_libs/lib.pyx [hunk 7]
```diff
@@ -477,7 +477,7 @@ def maybe_indices_to_slice(ndarray[int64_t] indices, int max_len):
 def maybe_booleans_to_slice(ndarray[uint8_t] mask):
     cdef:
         Py_ssize_t i, n = len(mask)
-        Py_ssize_t start, end
+        Py_ssize_t start = 0, end = 0
```
Comment (pilkibun) [inline pandas/_libs/lib.pyx]: This is an awkward little state machine. By my code analysis, this is a false positive. both will either not be referenced or be initialized first.

File: pandas/_libs/lib.pyx [hunk 8]
```diff
@@ -1631,7 +1631,7 @@ def is_datetime_with_singletz_array(values: ndarray) -> bool:
     Doesn't check values are datetime-like types.
     """
     cdef:
-        Py_ssize_t i, j, n = len(values)
+        Py_ssize_t i = 0, j, n = len(values)
```
Comment (pilkibun) [inline pandas/_libs/lib.pyx]: real bug. if n ==0, `range` falls through and `i` is referenced uninit...

File: pandas/_libs/hashtable.pxd (hunk 2 above)
Comment (WillAyd) [reply pandas/_libs/hashtable.pxd]: Your comment about `ssize_t` would apply generally, but most of this code operates in the context of Python which made the design decision to use the signed `Py_size_t` for indexing instead. So if this touches anything for indexing (which I think it does) will want to keep that consistent

File: pandas/_libs/lib.pyx [hunk 9]
```diff
@@ -1913,8 +1913,8 @@ def maybe_convert_objects(ndarray[object] objects, bint try_float=0,
         ndarray[int64_t] ints
         ndarray[uint64_t] uints
         ndarray[uint8_t] bools
-        ndarray[int64_t] idatetimes
-        ndarray[int64_t] itimedeltas
+        int64_t[::1]  idatetimes
+        int64_t[::1] itimedeltas
```
Comment (pilkibun) [inline pandas/_libs/lib.pyx]: compiler compalins about dim_info[0].strided being possibly uninited. I don't fully understand why, but using memory views generates different code and should be interchangeable.

File: pandas/_libs/src/parser/tokenizer.c [hunk 10]
```diff
@@ -424,13 +424,13 @@ static void append_warning(parser_t *self, const char *msg) {
 
     if (self->warn_msg == NULL) {
         self->warn_msg = (char *)malloc(length + 1);
-        strncpy(self->warn_msg, msg, strlen(msg) + 1);
+        snprintf(self->warn_msg, length + 1, "%s", msg);
```
Comment (pilkibun) [inline pandas/_libs/src/parser/tokenizer.c]: 1. https://stackoverflow.com/a/56782476
2. cpplint complains about using bare strcpy. 

snprintf is fine with both.

File: pandas/_libs/src/parser/tokenizer.c [hunk 11]
```diff
@@ -1815,7 +1816,7 @@ double round_trip(const char *p, char **q, char decimal, char sci, char tsep,
     double r = PyOS_string_to_double(p, q, 0);
     if (maybe_int != NULL) *maybe_int = 0;
     if (PyErr_Occurred() != NULL) *error = -1;
-    else if (r == Py_HUGE_VAL) *error = Py_HUGE_VAL;
+    else if (r == Py_HUGE_VAL) *error = (int)Py_HUGE_VAL;
```
Comment (pilkibun) [inline pandas/_libs/src/parser/tokenizer.c]: awkward code. Py_HUGE_VAL is a float and is used as an error code with  an implicit cast. made it explicit to silence the warning.

File: pandas/_libs/groupby_helper.pxi.in (hunk 1 above)
Comment (WillAyd) [reply pandas/_libs/groupby_helper.pxi.in]: Let's leave this out of this PR for now. To Jeff's point we should be consistent in how we clean up this warning which will require touching quite a few areas. Doing this properly will require proper templating and/or use of fused types

File: pandas/_libs/hashtable.pxd (hunk 2 above)
Comment (pilkibun) [reply pandas/_libs/hashtable.pxd]: `Py_size_t` is not even defined and  Python actually made the design choice to use Py_ssize_t everywhere.

If you mean use `Py_ssize_t` instead of `ssize_t `, ok, but that's a different issue.

Comment (WillAyd) [reply pandas/_libs/hashtable.pxd]: Yes

File: pandas/_libs/groupby_helper.pxi.in (hunk 1 above)
Comment (pilkibun) [reply pandas/_libs/groupby_helper.pxi.in]: no prob. updated.

Comment (pilkibun) [reply pandas/_libs/groupby_helper.pxi.in]: Just out of curioisity, how would fused types help and/or what is improper with this template? it's important, because interval.pyx has similar problem (warnings due to) with fused types covering both signed and unsigned types, and code that does comparisons. If you don't check by name, what would be a cleaner way?

Comment (WillAyd) [reply pandas/_libs/groupby_helper.pxi.in]: I am not discarding the approach just yet but whatever we do should be consistent. If you look in this module there are multiple ways of handling this (see line 391 for an alternate approach) and other modules may have their own handling

So best for a separate comprehensive PR to address and standardize

File: pandas/_libs/src/parser/tokenizer.c (hunk 11 above)
Comment (WillAyd) [reply pandas/_libs/src/parser/tokenizer.c]: Good find. Does the caller even use this?

File: pandas/_libs/lib.pyx (hunk 9 above)
Comment (WillAyd) [reply pandas/_libs/lib.pyx]: Makes sense conceptually. @jbrockmendel has good experience in converting between these though so may have insight to offer

File: pandas/_libs/lib.pyx (hunk 8 above)
Comment (WillAyd) [reply pandas/_libs/lib.pyx]: Wouldn't the `n == 0` check on the subsequent line which returns prevent this?

File: pandas/_libs/lib.pyx (hunk 7 above)
Comment (WillAyd) [reply pandas/_libs/lib.pyx]: Makes sense

File: pandas/_libs/index.pyx (hunk 6 above)
Comment (WillAyd) [reply pandas/_libs/index.pyx]: So this would fail if `len(values) == 0` right? Do you see a code sample to actually trigger that?

File: pandas/_libs/hashtable_func_helper.pxi.in (hunk 5 above)
Comment (WillAyd) [reply pandas/_libs/hashtable_func_helper.pxi.in]: Makes sense

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/_libs/groupby_helper.pxi.in (hunk 1 above)
Comment (pilkibun) [reply pandas/_libs/groupby_helper.pxi.in]: typo? line 391 doesn't seem relevant.

File: pandas/_libs/index.pyx (hunk 6 above)
Comment (pilkibun) [reply pandas/_libs/index.pyx]: quite possibly this is meant to be "caller checks". But as an independent module which that's out of the compiler's view.

File: pandas/_libs/lib.pyx (hunk 8 above)
Comment (pilkibun) [reply pandas/_libs/lib.pyx]: Yes. false positive then.

File: pandas/_libs/src/parser/tokenizer.c (hunk 11 above)
Comment (pilkibun) [reply pandas/_libs/src/parser/tokenizer.c]: DIdn't check. Probably err !=0 at least.

Comment (WillAyd) [reply pandas/_libs/src/parser/tokenizer.c]: No worries. Makes sense to suppress build warning though as you can tell there are a lot of opportunities to really clean some of this up. If you say any can certainly push PRs

Resolving this one for now

File: pandas/_libs/index.pyx (hunk 6 above)
Comment (WillAyd) [reply pandas/_libs/index.pyx]: Yea sure. Similar comment though - good to suppress compiler warning but if there's an actual error here (which I think would still happen without the warning) would be good to address

File: pandas/_libs/groupby_helper.pxi.in (hunk 1 above)
Comment (WillAyd) [reply pandas/_libs/groupby_helper.pxi.in]: https://github.com/pandas-dev/pandas/blob/355e322b5ed4a31c9e1cab9652510e85a54f23a3/pandas/_libs/groupby_helper.pxi.in#L391

Comment (pilkibun) [reply pandas/_libs/groupby_helper.pxi.in]: ah. different commit. I removed these and similar change.

File: pandas/_libs/hashtable.pxd (hunk 2 above)
Comment (pilkibun) [reply pandas/_libs/hashtable.pxd]: done

File: pandas/_libs/index.pyx (hunk 6 above)
Comment (pilkibun) [reply pandas/_libs/index.pyx]: if values is empty, the loop falls through, `util.get_value_at(values, mid)` gets called when values is empty, which calls validate_indexer which raises. I think is ok.

**Comment by pilkibun** [comment]
Green and I think I've addressed all the comments.

File: pandas/_libs/src/parser/tokenizer.c [hunk 12]
```diff
@@ -1433,13 +1433,14 @@ PANDAS_INLINE void uppercase(char *p) {
 int to_boolean(const char *item, uint8_t *val) {
     char *tmp;
     int i, status = 0;
-    int bufsize = sizeof(char) * (strlen(item) + 1);
+    size_t length0 = (strlen(item) + 1);
+    int bufsize = sizeof(char) * length0;
```
Comment (WillAyd) [inline pandas/_libs/src/parser/tokenizer.c]: Can just get rid of `sizeof(char)` here as C standard guarantees this to be 1

File: pandas/_libs/hashtable.pyx (hunk 4 above)
Comment (WillAyd) [reply pandas/_libs/hashtable.pyx]: I think Py_ssize_t here as well

**Review by WillAyd — CHANGES_REQUESTED** [review]
minor nits other lgtm

File: pandas/_libs/groupby_helper.pxi.in (hunk 1 above)
Comment (pilkibun) [reply pandas/_libs/groupby_helper.pxi.in]: You know what? This is really irritating.

My changes elicit a "we don’t do this anywhere else" response from jreback, and "whatever we do should be consistent" from you, when in the meanwhile https://github.com/pandas-dev/pandas/blob/355e322b5ed4a31c9e1cab9652510e85a54f23a3/pandas/_libs/groupby_helper.pxi.in#L56-L60

and https://github.com/pandas-dev/pandas/blob/355e322b5ed4a31c9e1cab9652510e85a54f23a3/pandas/_libs/groupby_helper.pxi.in#L113-L117

and 
https://github.com/pandas-dev/pandas/blob/355e322b5ed4a31c9e1cab9652510e85a54f23a3/pandas/_libs/groupby_helper.pxi.in#L212-L216

and multiple others. 

Why reject fixes for an actual problem, in the name of a code style that may or may not someday come to be? It'll probably be this very same thing, by the look of it.

File: pandas/_libs/src/parser/tokenizer.c (hunk 12 above)
Comment (pilkibun) [reply pandas/_libs/src/parser/tokenizer.c]: yeah

File: pandas/_libs/hashtable.pyx (hunk 4 above)
Comment (pilkibun) [reply pandas/_libs/hashtable.pyx]: done.

File: pandas/_libs/groupby_helper.pxi.in (hunk 1 above)
Comment (WillAyd) [reply pandas/_libs/groupby_helper.pxi.in]: Right those are inconsistent and none match what you have in this PR. I understand the frustration (I've introduced some of that inconsistency) but we are just trying to improve the code base

In any case certainly would be welcome the fix but separate PR as it needs a more comprehensive look on the commit and review side of things

File: pandas/_libs/src/parser/tokenizer.c [hunk 13]
```diff
@@ -1433,13 +1433,14 @@ PANDAS_INLINE void uppercase(char *p) {
 int to_boolean(const char *item, uint8_t *val) {
     char *tmp;
     int i, status = 0;
-    int bufsize = sizeof(char) * (strlen(item) + 1);
+    size_t length0 = (strlen(item) + 1);
+    int bufsize = length0;
```
Comment (WillAyd) [inline pandas/_libs/src/parser/tokenizer.c]: This doesn't introduce any new warnings right? Implicitly casting size_t to int?

**Review by WillAyd — APPROVED** [review]
Think this is good. Maybe one comment to look at

File: pandas/_libs/groupby_helper.pxi.in (hunk 1 above)
Comment (pilkibun) [reply pandas/_libs/groupby_helper.pxi.in]: argh. moving on.

File: pandas/_libs/src/parser/tokenizer.c (hunk 13 above)
Comment (pilkibun) [reply pandas/_libs/src/parser/tokenizer.c]: checked now, looks ok. anyway the original line's RHS looks like it was size_t as well, no?

File: pandas/_libs/index.pyx (hunk 6 above)
Comment (jbrockmendel) [reply pandas/_libs/index.pyx]: > which raises

cython code can have weird corner cases where exceptions get ignored.  Adding a test for this case seems worthwhile.

File: pandas/_libs/lib.pyx (hunk 9 above)
Comment (jbrockmendel) [reply pandas/_libs/lib.pyx]: IIRC this function won't return idatetimes or itimedeltas to the user, so this should be fine.  Is there a particular reason for using `::1` instead of just `:`?  We generally do the latter.

File: pandas/_libs/src/parser/tokenizer.c [hunk 14]
```diff
@@ -1433,13 +1433,14 @@ PANDAS_INLINE void uppercase(char *p) {
...
 
     static const char *tstrs[1] = {"TRUE"};
     static const char *fstrs[1] = {"FALSE"};
 
     tmp = malloc(bufsize);
-    strncpy(tmp, item, bufsize);
+    snprintf(tmp,  length0, "%s", item);
```
Comment (jbrockmendel) [inline pandas/_libs/src/parser/tokenizer.c]: For my own edification: is there any potential performance difference here?

File: pandas/_libs/src/parser/tokenizer.c (hunk 11 above)
Comment (jbrockmendel) [reply pandas/_libs/src/parser/tokenizer.c]: @pilkibun this part of the code gets less maintenance attention than other parts; is likely to have developed unused/unreachable branches.  Any of those you can identify and remove (separate PR(s)) will be welcome.

**Comment by jbrockmendel** [comment]
> np_datetime.c and np_datetime_strings.c. These are linked into 15+ cython modules as extra object files, but they '#include <datetime.h>`, which expects only standalone modules, who properly init to do it. Kinda painful to fix, though it means each function is compiled and kept in memory 15 times.

One more for my edification.  Is this because they are specified in setup.py, or because of cimports from np_datetime.pyx, or [...]?  How costly is the "compile and kept in memory 15 times"?

File: pandas/_libs/src/parser/tokenizer.c (hunk 11 above)
Comment (pilkibun) [reply pandas/_libs/src/parser/tokenizer.c]: i ph33r the parser code.

File: pandas/_libs/index.pyx (hunk 6 above)
Comment (pilkibun) [reply pandas/_libs/index.pyx]: interesting. example?

File: pandas/_libs/lib.pyx (hunk 9 above)
Comment (jreback) [reply pandas/_libs/lib.pyx]: @pilkibun if you'd change this to ``[:]`` otherwise this PR lgtm.

File: pandas/_libs/src/parser/tokenizer.c (hunk 14 above)
Comment (pilkibun) [reply pandas/_libs/src/parser/tokenizer.c]: potential? sure. effect size? not sure.  I'm not about to benchmark snprintf vs. strcpy for this silliness.
afaict, those are our choices:

- strncpy causes compiler warning, because this is not how it's meant to be used
- strcpy is flagged by cpplint, and there's no per-line ignore mechanism afaict. can disable printf checking for the whole file using a CPPLINT.cfg file.
- snprintf involves some overhead. at least "parsing" "%s", some input validation and an additional function call to strcpy. call it a few us per call.

File: pandas/_libs/lib.pyx (hunk 9 above)
Comment (pilkibun) [reply pandas/_libs/lib.pyx]: That effectively means we expect `np.empty` to return a c-contig array. I'll lose it.

**Comment by pilkibun** [comment]
> One more for my edification. Is this because they are specified in setup.py, or because of cimports 
> from np_datetime.pyx, or [...]? How costly is the "compile and kept in memory 15 times"?

I came across your comment in the file:
https://github.com/pandas-dev/pandas/blob/02b552d280f917c1a181fcfd956a2d01f0bd5462/pandas/_libs/tslibs/src/datetime/np_datetime.c#L339-L341

The issue is that it's supposed to be an independent cython/cpython module, and imported by other cython modules. Then it could call IMPORT, and a few other warnings would go away.

Instead these files are compiled into object files, and linked into each module that uses them. there's no crash because each module does call IMPORT, but the compiler can't see that.

I don't think compilation is an issue, because:
1. its not a large file
2. cython generates monstrosities, so any true c pales
3. On most dev machine, you have ccache, so (maybe, possibly) the compilation result is cached
4. depending on how  distutils compiles things, it's linked 15 times, but the object is the same dependency for all of them in the Makefile sense, so compiled only once/

as for memory waste it's 2019, so no not an issue.

In summary: more of a smell than a perf/resource issue.

File: pandas/_libs/index.pyx (hunk 6 above)
Comment (pilkibun) [reply pandas/_libs/index.pyx]: huh. I think you've nailed something here. So it's not exactly weird corner  cases, but cdef function which return exception as if they were python.
https://cython.readthedocs.io/en/latest/src/userguide/language_basics.html#error-return-values

`validate_indexer` and `get_value_at` are both cdef. validate_indexer has ``except -1`, so any exception translates into a -1 retval. But the caller `get_value_at` doesn't check that and uses -1 to reference into the the empty array.

OTOH, the index constructor refuses to create empty indexes, so an end-to-end test does not seem possible. and  `get_value_at`  is used all over the place.

Hmm. will have to think about this.

Comment (pilkibun) [reply pandas/_libs/index.pyx]:
```python
In [18]: %%cython
    ...: 
    ...: cdef int foo() except -1:
    ...:     1/0
    ...:     cdef:
    ...:         int mid = 42
    ...:     return mid
    ...:     
    ...: cdef int bar():
    ...:     res = foo()
    ...:     return res
    ...:     
    ...: def baz():
    ...:     return bar()

In [19]: baz()
---------------------------------------------------------------------------
ZeroDivisionError                         Traceback (most recent call last)
_cython_magic_a591cd36cd8e49116d6e370949ee24c2.pyx in _cython_magic_a591cd36cd8e49116d6e370949ee24c2.foo()

ZeroDivisionError: float division
Exception ignored in: '_cython_magic_a591cd36cd8e49116d6e370949ee24c2.bar'
Traceback (most recent call last):
  File "_cython_magic_a591cd36cd8e49116d6e370949ee24c2.pyx", line 3, in _cython_magic_a591cd36cd8e49116d6e370949ee24c2.foo
ZeroDivisionError: float division
Out[19]: 0
```

very interesting bug.

File: pandas/_libs/index.pyx [hunk 15]
```diff
@@ -355,7 +355,7 @@ cdef Py_ssize_t _bin_search(ndarray values, object val) except -1:
         Py_ssize_t mid = 0, lo = 0, hi = len(values) - 1
         object pval
 
-    if hi >= 0 and val > util.get_value_at(values, hi):
+    if hi == 0 or (hi > 0 and val > util.get_value_at(values, hi)):
```
Comment (jreback) [inline pandas/_libs/index.pyx]: is there any user code that hits this? e.g. is there some example that failed before and now works? or just an untested corner case?

**Review by jreback — APPROVED** [review]

File: pandas/_libs/index.pyx (hunk 15 above)
Comment (pilkibun) [reply pandas/_libs/index.pyx]: the only caller is INdexEngine.get_loc, and Index constructor (I imagine all of them) will refuse to create an empty index. So, this cannot be triggered. Reviewers pressed, so I responded.

**Comment by jreback** [comment]
ok thanks
